### ISSUE TITLE
Add ValueKind property in System.Text.Json.Nodes.JsonNode class

### ISSUE THREAD

**Issue by kingcean** [dotnet/runtime#53406]
## Background and Motivation

`System.Text.Json.Node.JsonNode` has provided a way to `as` `JsonObject`, `JsonArray` and `JsonValue`, but has no way to get its value kind directly.

## Proposed API

```diff
namespace System.Text.Json.Nodes
{
    public abstract class JsonNode : IDynamicMetaObjectProvider {
+        public abstract JsonValueKind ValueKind { get; }
    }
}
```

## Usage Examples

``` C#
var json = JsonNode.Parse("{ \"name\": \"Kingcean\" }");
Assert.AreEqual(JsonValueKind.Object, json.ValueKind );
Assert.AreEqual(JsonValueKind.String, json["name"].ValueKind ;
```

## Implementation for Reference

```C#
// namespace System.Text.Json.Nodes
public abstract class JsonNode
{
    public abstract JsonValueKind ValueKind { get; }
    // ...
}
public sealed class JsonArray : JsonNode, IList<JsonNode?>, ICollection<JsonNode?>, IEnumerable<JsonNode?>, IEnumerable
{
    public override JsonValueKind ValueKind => JsonValueKind.Array;
    // ...
}
public sealed class JsonObject : JsonNode, IDictionary<string, JsonNode?>, ICollection<KeyValuePair<string, JsonNode?>>, IEnumerable<KeyValuePair<string, JsonNode?>>, IEnumerable
{
    public override JsonValueKind ValueKind => JsonValueKind.Object;
    // ...
}
internal abstract partial class JsonValue<TValue> : JsonValue
{
    public override JsonValueKind ValueKind
    {
        get
        {
            if (_value is null) return JsonValueKind.Null;
            if (_value is JsonElement element) return element.ValueKind;
            if (_value is string || _value is Guid || _value is DateTime || _value is DateTimeOffset) return JsonValueKind.String;
            if (_value is bool b) return b ? JsonValueKind.True : JsonValueKind.False;
            if (_value is int || _value is long || _value is uint || _value is ulong || _value is short || _value is ushort || _value is double || _value is float || _value is decimal || _value is byte ||  || _value is sbyte) return JsonValueKind.Number;
            throw new NotSupportedException();
        }
    }
    // ...
}
```

**Comment by kevinchalet** [comment]
Having such a property would be extremely useful! 👏🏻 

/cc @steveharter

**Comment by kevinchalet** [comment]
As mentioned in https://github.com/dotnet/runtime/issues/52611, this property would be very useful in OpenIddict. Any chance it could be discussed for 6.0?

**Comment by steveharter** [comment]
The main reason that this isn't supported is that a `JsonValue` in "edit mode" can be assigned to any CLR type - a POCO, a Dictionary, an Enum, etc and in JSON they may be serialized in an arbitrary format depending on the converter. In "read mode" (after a `Parse()`) the underlying `JsonElement.ValueKind` can be used.

However, I did create a `JsonNode` extension sample for `ValueKind` in https://github.com/dotnet/runtime/issues/55827#issuecomment-889443545 that handles this pretty well, but is not cheap when in "edit" mode and these types of values are assigned to a `JsonValue`. I plan on formalizing the sample soon (in STJ tests) and depending on feedback may make it into 7.0.

Also, note that `JsonNode.GetValue<object>()` can be called to get the underlying value. This may be useful in certain scenarios, like `ValueKind`,  where inspection of the value is necessary. I'm also updating the [doc ](https://github.com/dotnet/runtime/pull/56639) to make this more clear.

Closing as it should be addressed by https://github.com/dotnet/runtime/issues/55827.
